README.md
# NAME

Badge::Depot::Plugin::Coverage - Code coverage plugin for Badge::Depot

# VERSION

Version 0.0102, released 2020-12-26.

# SYNOPSIS

Used standalone:

    use Badge::Depot::Plugin::Coverage;

    my $badge = Badge::Depot::Plugin::Coverage->new(coverage => 87);

    print $badge->to_html;

Used together with [Pod::Weaver::Section::Badges](https://metacpan.org/pod/Pod::Weaver::Section::Badges), in weaver.ini:

    [Badges]
    ; other settings
    badge = Coverage

# DESCRIPTION

This [Badge::Depot](https://metacpan.org/pod/Badge::Depot) badge is meant to be used together with [Dist::Zilla::App::Command::coverh](https://metacpan.org/pod/Dist::Zilla::App::Command::coverh) (or standalone, as per the synopsis) and creates a coverage badge:

![Coverage 87%](https://img.shields.io/badge/coverage-87%25-orange.svg)

# ATTRIBUTES

## coverage

Set the code coverage percentage manually. Should only be used when [Dist::Zilla::App::Command::coverh](https://metacpan.org/pod/Dist::Zilla::App::Command::coverh) is **not** used.

## custom\_image\_url

Default: `https://img.shields.io/badge/%s-%s-%s.svg`

Override the default image url. It is expected to have three `sprintf` placeholders: Text, coverage percentage and color.

## max\_age

Default: `60`

When used together with [Dist::Zilla::App::Command::coverh](https://metacpan.org/pod/Dist::Zilla::App::Command::coverh), only include the badge if the latest coverage run was less than `max_age` minutes ago.

## text

Default: `coverage`

Set a different coverage percentage. The percentage sign will be appended.

# SEE ALSO

- [Badge::Depot](https://metacpan.org/pod/Badge::Depot)
- [Task::Badge::Depot](https://metacpan.org/pod/Task::Badge::Depot)

# SOURCE

[https://github.com/Csson/p5-Badge-Depot-Plugin-Coverage](https://github.com/Csson/p5-Badge-Depot-Plugin-Coverage)

# HOMEPAGE

[https://metacpan.org/release/Badge-Depot-Plugin-Coverage](https://metacpan.org/release/Badge-Depot-Plugin-Coverage)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Badge-Depot-Plugin-Coverage

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0102 2020-12-26T21:35:37Z UTC
 [Bug Fixes]
 - Url-encode the percentage sign

 [Stats & Dependency Changes]
 - 3 files changed, 55 insertions(+), 6 deletions(-)
 - (run req) ~ DateTime (any) --> 1.05
 - (run req) ~ JSON::MaybeXS (any) --> 1.003003
 - (run req) ~ Moose (any) --> 2.1803
 - (run req) ~ MooseX::AttributeShortcuts (any) --> 0.028
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Types::Standard (any) --> 1.000006
 - (run req) ~ namespace::autoclean (any) --> 0.22
 - (dev req) + Dist::Zilla 6.015
 - (dev req) +
   Dist::Zilla::Plugin::Author::CSSON::GithubActions::Workflow::TestWithMakefile
   0.0106
 - (dev req) + Dist::Zilla::Plugin::Git 2.046
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale 0.057
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   (any) --> 0.009
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
   --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.011
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.170880
 - (dev req) ~ Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013 -->
   0.014
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.035
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.004003
 - (dev req) ~ Dist::Zilla::Plugin::NextRelease::Grouped (any) --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.163250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.029
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.058
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOF (any) --> 0.0501
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOL (any) --> 0.18
 - (dev req) ~ Dist::Zilla::Plugin::Test::NoTabs (any) --> 0.15
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.027
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version (any) --> 1.09
 - (dev req) ~ Test::NoTabs (any) --> 1.4
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::Test::Kwalitee::Extra
 - (dev req) - Dist::Zilla::Plugin::TravisYML
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) ~ Dist::Iller 0.1406 --> 0.1409
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0307 --> 0.0327

0.0101 2016-02-17T19:13:34Z UTC
 [Bug Fixes]
 - Make it work on 5.10, for real.

 [Stats & Dependency Changes]
 - 2 files changed, 9 insertions(+), 3 deletions(-)
 - (run req) ~ perl 5.010001 --> 5.010000
 - (test req) + IO::Handle (any)
 - (test req) + IPC::Open3 (any)
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   (any)
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped (any)
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::Compile (any)
 - (dev req) + Dist::Zilla::Plugin::Test::EOF (any)
 - (dev req) + Dist::Zilla::Plugin::Test::EOL (any)
 - (dev req) + Dist::Zilla::Plugin::Test::Kwalitee::Extra (any)
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs (any)
 - (dev req) + Dist::Zilla::Plugin::Test::Version (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs (any)
 - (dev req) - Dist::Zilla::Plugin::CheckExtraTests
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Pod::Weaver::Section::Badges
 - (dev sug) + Dist::Iller 0.1406
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0307

0.0100 2016-01-19T00:24:23Z UTC
 - Initial release


lib/Badge/Depot/Plugin/Coverage.pm
use 5.10.0;
use strict;
use warnings;
use feature qw/say/;

package Badge::Depot::Plugin::Coverage;

use Moose;
use namespace::autoclean;
use MooseX::AttributeShortcuts;
use Types::Standard qw/Str Bool Num/;
use Types::URI qw/Uri/;
use JSON::MaybeXS 'decode_json';
use Path::Tiny;
use DateTime;
use DateTime::Format::RFC3339;
with 'Badge::Depot';

# ABSTRACT: Code coverage plugin for Badge::Depot
# AUTHORITY
our $VERSION = '0.0103';

has coverage => (
    is => 'ro',
    isa => Num,
    required => 0,
    predicate => 1,
);
has custom_image_url => (
    is => 'ro',
    isa => Uri,
    coerce => 1,
    default => 'https://img.shields.io/badge/%s-%s-%s.svg',
);
has text => (
    is => 'ro',
    isa => Str,
    default => 'coverage',
);
has max_age => (
    is => 'ro',
    isa => 'Int',
    default => '60',
    documentation => q{Include coverage badge if the latest cover run was done less than 'max_age' minutes ago},
);


sub BUILD {
    my $self = shift;

    my $coverage = $self->determine_coverage;
    if($coverage == -1) {
        $self->log('! Could not determine coverage, skips Coverage badge');
        return;
    }

    $coverage = sprintf '%.1f', $coverage;
    my $color = $self->determine_color($coverage);
    $self->image_url(sprintf $self->custom_image_url, $self->text, $coverage.'%25', $color);
    $self->image_alt(sprintf '%s %s', $self->text, $coverage.'%');
    $self->log("Adds coverage badge ($coverage%)");
}


sub determine_coverage {
    my $self = shift;
    return $self->coverage if $self->has_coverage;

    my $log_file = path(qw/. .coverhistory.json/);

    if(!$log_file->exists) {
        $self->log('! Could not find cover history file (.coverhistory.json), no coverage to add');
        return -1;
    }

    my $history = decode_json($log_file->slurp);

    if(!scalar @$history) {
        $self->log('! No coverage history in .coverhistory.json, nothing to add');
        return -1;
    }

    my $summary = $history->[-1];
    my $created_at = DateTime::Format::RFC3339->parse_datetime($summary->{'created_at'});
    my $since = DateTime->now - $created_at;

    my $minutes = $since->years * 1440 * 365
                + $since->months * 1440 * 30
                + $since->days * 1440
                + $since->hours * 60
                + $since->minutes
                + $since->seconds / 60;

    if($minutes > $self->max_age) {
        $self->log("! Latest coverage was run @{[ sprintf '%.1f', $minutes ]} minutes ago, maximum allowed age is @{[ $self->max_age ]} minutes, wont add coverage");
        return -1;
    }
    return $summary->{'total'};
}
sub determine_color {
    my $self = shift;
    my $coverage = shift;

    return $coverage < 75 ? 'red'
         : $coverage < 90 ? 'orange'
         : $coverage < 100 ? 'yellow'
         : $coverage == 100 ? 'brightgreen'
         :                     'blue'
         ;
}
sub log {
    my $self = shift;
    my $text = shift;

    say "[Badge/Coverage] $text";

};

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=head1 SYNOPSIS

Used standalone:

    use Badge::Depot::Plugin::Coverage;

    my $badge = Badge::Depot::Plugin::Coverage->new(coverage => 87);

    print $badge->to_html;

Used together with L<Pod::Weaver::Section::Badges>, in weaver.ini:

    [Badges]
    ; other settings
    badge = Coverage

=head1 DESCRIPTION

This L<Badge::Depot> badge is meant to be used together with L<Dist::Zilla::App::Command::coverh> (or standalone, as per the synopsis) and creates a coverage badge:

=for HTML <p><img src="https://img.shields.io/badge/coverage-87%25-orange.svg" /></p>

=for markdown ![Coverage 87%](https://img.shields.io/badge/coverage-87%25-orange.svg)

=head1 ATTRIBUTES

=head2 coverage

Set the code coverage percentage manually. Should only be used when L<Dist::Zilla::App::Command::coverh> is B<not> used.

=head2 custom_image_url

Default: C<https://img.shields.io/badge/%s-%s-%s.svg>

Override the default image url. It is expected to have three C<sprintf> placeholders: Text, coverage percentage and color.

=head2 max_age

Default: C<60>

When used together with L<Dist::Zilla::App::Command::coverh>, only include the badge if the latest coverage run was less than C<max_age> minutes ago.

=head2 text

Default: C<coverage>

Set a different coverage percentage. The percentage sign will be appended.

=head1 SEE ALSO

=for :list
* L<Badge::Depot>
* L<Task::Badge::Depot>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Code coverage plugin for Badge::Depot",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Badge-Depot-Plugin-Coverage",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Badge::Depot::Plugin::Coverage",
  "PREREQ_PM" => {
    "Badge::Depot" => 0,
    "DateTime" => "1.05",
    "DateTime::Format::RFC3339" => 0,
    "JSON::MaybeXS" => "1.003003",
    "Moose" => "2.1803",
    "MooseX::AttributeShortcuts" => "0.028",
    "Path::Tiny" => "0.072",
    "Types::Standard" => "1.000006",
    "Types::URI" => 0,
    "namespace::autoclean" => "0.22"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0103",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Badge::Depot" => 0,
  "DateTime" => "1.05",
  "DateTime::Format::RFC3339" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeXS" => "1.003003",
  "Moose" => "2.1803",
  "MooseX::AttributeShortcuts" => "0.028",
  "Path::Tiny" => "0.072",
  "Test::More" => "0.96",
  "Types::Standard" => "1.000006",
  "Types::URI" => 0,
  "namespace::autoclean" => "0.22"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
